observatory-dags/observatory/dags/telescopes/crossref_events.py

```diff
@@ -0,0 +1,388 @@
+
+        :return:
+        """
+        all_results = []
+
+        if CrossrefEventsTelescope.DOWNLOAD_MODE == 'parallel':
+            no_workers = CrossrefEventsTelescope.MAX_PROCESSES
+            logging.info(f'Downloading events with parallel method, no. workers: {no_workers}')
+
+        else:
+            logging.info('Downloading events with sequential method')
+            no_workers = 1
+
+        with ThreadPoolExecutor(max_workers=no_workers) as executor:
+            futures = []
+            # select minimum, either no of batches or no of workers
+            for i in range(min(no_workers, len(self.urls))):
+                futures.append(executor.submit(download_events_batch, self, i))
+            for future in as_completed(futures):
+                batch_result = future.result()
+                all_results += batch_result
+
+        # get paths of failed request attempts
+        failed_files = [result[0] for result in all_results if not result[1]]
+        if failed_files:
+            raise AirflowException(f'Max retries exceeded for file(s): '
+                                   f'{", ".join(failed_files)}, saved cursor to corresponding file.')
+
+        batch_files = [result[0] for result in all_results]
+        # combine all event types for all batches to one file
+        continue_dag = False
+        with open(self.download_path, 'w') as fout, fileinput.input(batch_files) as fin:
+            for line in fin:
+                if line != '[]\n':
+                    fout.write(line)
+                    continue_dag = True
+
+        # only continue if file is not empty
+        return continue_dag
+
+    def transform(self):
+        """ Transforms the crossref events release. The download file contains multiple lists, one for each request,
+            and each list contains multiple events. Each event is transformed so that the field names do not contain
+            '-' and have a valid timestamp at 'occurred_at'. The events are written out individually and separated by a newline.
+            :return: None
+            """
+        # load and transform events
+        events = []
+        with open(self.download_path, 'r') as f:
+            for line in f:
+                for event in json.loads(line):
+                    try:
+                        pendulum.parse(event['occurred_at'])
+                    except ValueError:
+                        event['occurred_at'] = "0001-01-01T00:00:00Z"
+                    transformed_event = change_keys(event, convert)
+                    events.append(transformed_event)
+
+        list_to_jsonl_gz(self.transform_path, events)
+
+
+class CrossrefEventsTelescope(StreamTelescope):
+    """ Crossref Events telescope """
+
+    DAG_ID = 'crossref_events'
+
+    MAILTO = 'aniek.roelofs@curtin.edu.au'
+    DOWNLOAD_MODE = 'parallel'  # Valid options: ['sequential', 'parallel']
+    MAX_PROCESSES = 21  # <ax processes based on 7 days x 3 url categories
```

Comment:
So that the email address used with Crossref Events can be customised, it would also be good if `mail_to` was a mandatory argument in the constructor. In the file `observatory-dags/observatory/dags/dags/crossref_events.py` you would then fetch the email address from an Airflow Connection.

It would also be good if `download_mode` and `max_processes` were keyword arguments in the constructor so that these can be customised when the telescope is created.

You might have to pass some of these values to the release object.

Reply:
Thanks, that makes sense. I've added them to the telescope and release objects.